src/reducers/join.ts

```diff
@@ -0,0 +1,54 @@
+import _ from 'lodash'
+import existingThoughtChange from './existingThoughtChange'
+import { getAllChildren, getChildPath, getNextRank, getPrevRank } from '../selectors'
+import { SimplePath } from '../types'
+import { head, parentOf, pathToContext, reducerFlow } from '../util'
+import { State } from '../util/initialState'
+import existingThoughtMove from './existingThoughtMove'
+import existingThoughtDelete from './existingThoughtDelete'
+
+/** Join two or more thoughts split by spaces. */
+const join = (state: State) => {
+
+  const { cursor } = state
+
+  if (!cursor) return state
+
+  const contextRanked = pathToContext(parentOf(cursor))
+  const contextChildren = getAllChildren(state, contextRanked)
+  const { value, rank } = head(cursor)
+  const siblings = contextChildren.filter(child => child.value !== value && child.rank !== rank)
+
+  const actions = siblings.reduce((acc, sibling, index) => {
+
+    const pathToSibling = [...parentOf(cursor), sibling]
+    const children = getAllChildren(state, pathToContext(pathToSibling))
+
+    return [...acc, ...children.map(child => {
+      const oldPath = getChildPath(state, child, pathToSibling as SimplePath)
+      const newPath = [...cursor, { ...child, rank: index === 1 ? getNextRank(state, pathToContext(pathToSibling)) : getPrevRank(state, pathToContext(pathToSibling)) }]
```

Comment:
Since all the merged children are added to the end of the context, we should be able to call `getNextRank` once outside the loop and then increment by 1 for each subsequent sibling.

e.g.

- a
  - m
    - w (rank 0)
    - x (rank 1)
  - n
    - y (rank 0)
  - o
    - z (rank 0)

`getNextRank(state, [a', 'm'])` would return 1, and we could increment the rank of each merged child:

- a
  - m
    - w (rank 0)
    - x (rank 1)
    - y (rank 2)
    - z (rank 3)

Reply:
> Since all the merged children are added to the end of the context, we should be able to call `getNextRank` once outside the loop and then increment by 1 for each subsequent sibling.

I didn't know that the merged children were supposed to be added to the end of the context because it's not stated anywhere in the [issue description](https://github.com/cybersemics/em/issues/851#issue-688219266).
So my assumption was that we'd keep the order of children. For instance, running a join command on this structure

> * a
>   
>   * m
>     
>     * w (rank 0)
>     * x (rank 1)
>   * n (cursor)
>     
>     * y (rank 0)
>   * o
>     
>     * z (rank 0)

should result in

> * a
>   
>   * m n o (cursor)
>     * w
>     * x
>     * y
>     * z

If we append all the children to the end, however, we'd get something like this

> * a
>   
>   * m n o (cursor)
>     * y
>     * w
>     * x
>     * z


@raineorshine Please confirm which one is correct. And let's add the same to the issue description.

Reply:
Sorry for the confusion! Your original interpretation is correct. By "end of context" I didn't mean append without changing their ranks. I was trying to refer to what you were already doing.

The suggestion is related to efficiency rather than functionality. Do you think that an autoincrement would work and allow us to avoid calling `getNextRank` multiple times?

Reply:
I agree. In fact, the previous approach wasn't fully correct as it resulted in duplicate ranks. 
Though the `incrementing option` does the job, the code doesn't look very efficient now. So I'd like to hear your thoughts on that.

Reply:
> Though the `incrementing option` does the job, the code doesn't look very efficient now. So I'd like to hear your thoughts on that.

It is actually more efficient than before since it doesn't call `getNextRank` every iteration. It can simplified though.

I think you can just use `nextRankMin + i`...?

I'm not understanding the purpose of `index < cursorIndex`. If you think it's still needed maybe you can help me understand.

Reply:
I'm not sure about that. With the current approach, we're updating (incrementing or decrementing) the rank of every child **except the cursor head** (`y` in the below example). 
> * a
>   * m
>     * w (rank 0)
>     * x (rank 1)
>   * n (cursor)
>     * y (rank 0)
>   * o
>     * z (rank 0)

The result looks like this 
> * a
>   * m n o
>     * w (rank -2)
>     * x (rank -1)
>     * y (rank 0)
>     * z (rank 1)


The notion behind this approach was that we don't update the ranks of any thought in the current cursor as that might result in some side-effects. 

However, if you think otherwise, we can update the ranks of all merged children.
The result of that approach might look something like this -

> * a
>   * m n o
>     * w (rank 1)
>     * x (rank 2)
>     * y (rank 3)
>     * z (rank 4)

Reply:
> The notion behind this approach was that we don't update the ranks of any thought in the current cursor as that might result in some side-effects.

I don't believe it's necessary. We can re-rank from the new first child, starting at `0`, with no side effects.